File: GO/day_11/part2.go
package day11

import (
	"fmt"

	"aoc2025/utils"
)

type state struct {
	key string
	f   bool
	d   bool
}

func Part2(f string) {
	data := utils.ReadStringLines(path[f])
	adj := createAdjMap(data)
	memo := make(map[state]int)

	var dfs func(string, bool, bool) int
	dfs = func(s string, f, d bool) int {
		if val, ok := memo[state{s, f, d}]; ok {
			return val
		}

		if s == "out" {
			if f && d {
				return 1
			}
			return 0
		}

		total := 0

		for _, v := range adj[s] {
			nf := f || v == "fft"
			nd := d || v == "dac"

			total += dfs(v, nf, nd)
		}

		memo[state{s, f, d}] = total

		return total
	}

	sum := dfs("svr", false, false)

	fmt.Println(sum)
}

File: JS/day_11/part2.js
import { readFile } from "node:fs/promises";

async function part2(path) {
  const file = await readFile(path, "utf8");
  const data = file.trim().split("\n");

  const adj = createAdj(data);
  const memo = new Map();

  function dfs(s, f, d) {
    const key = `${s}|${f}|${d}`;
    if (memo.has(key)) return memo.get(key);

    if (s === "out") {
      const res = f && d ? 1 : 0;
      memo.set(key, res);
      return res;
    }

    let total = 0;

    adj[s].forEach((el) => {
      const nf = f || el === "fft";
      const nd = d || el === "dac";
      total += dfs(el, nf, nd);
    });

    memo.set(key, total);
    return total;
  }

  const sum = dfs("svr", false, false);
  console.log("Day 11 | Part 2:", sum);
}

function createAdj(data) {
  const adj = {};

  data.forEach((el) => {
    const parts = el.split(":");
    adj[parts[0]] = parts[1].trim().split(" ");
  });

  return adj;
}

part2("input.txt", false, false);
